lrpar/cttests/src/calc_parse_param.test

```diff
@@ -0,0 +1,49 @@
+name: Test %parse-param using the calculator grammar
+yacckind: Original(YaccOriginalActionKind::UserAction)
+grammar: |
+    %start Expr
+    %actiontype Result<u64, ()>
+    %avoid_insert 'INT'
+    %parse-param <'a> {x: &'a mut int}
```

Comment:
I don't have a deep opinion here, but my gut reaction is that a) we should call it `%parse_param` (to match the `_` in `%avoid_insert) b) I think the `<...>` lifetime syntax makes sense c) the `{...}` parameter syntax makes a bit less sense (at first I thought it was action code until I read Bison's documentation). I wondered if either changing `{}` -> `()` would make sense, or maybe just drop the brackets altogether? i.e. either `%parse_param <'a> (x: &'a mut int)` or `%parse_param <'a> x: &'a mut int`.

Reply:
Agree, I'd only done that because thats the way bison did it -- matching rust syntax is probably a better option and less work.
thank you for the feedback.

Reply:
My pleasure -- I think this is a useful feature so I'm happy to work with you to get it in!

Reply:
I went with the latter, there is a (probably excessive) comment which gives a rationale for the choice. That wins out over the `(x: &'a mut int)` because rust types may contain `(int, int)` making it a somewhat painful exercise.

Reply:
I found the single variable case in https://github.com/softdevteam/grmtools/pull/214/commits/1eb94e0218344ecb394d3f8dfb4f572c315e6a69 to be clear, but the multi-variable case confusing. I reckon it probably needs to be a bit closer to Rust's syntax, so something like `%parse_param <'a> (x: int, y: int)` (or `%parse_namar <'a> x: int, y: int`) perhaps?

Reply:
I agree it isn't as nice, it strangely *is* using Rust syntax, but Rust's syntax isn't uniform between parameters and tuples, and in this case it's very much easier to use the tuple syntax.

Both of the above options are unfortunately going to force us to parse a good amount of rust arrays, tuples, and `<T1, T2>`  because we need to pull out `x`, and `y` to pass them down to the user actions from parse, so we'll need to know the variable names.

Now, because `fn parse<'a>(s:&'a str, (x, y) : (u32, u32)) {}` is valid syntax,
but `fn parse<'a>(s:&'a str, (x: u32, y: u32)){}` is not.

We'd have to either parse out x and y either to get the variable names, or to translate them into tuple decl syntax.

And when it comes to  `fn parse<'a>(s:&'a str, x: Foo<(u32, u32), u32>, y: u32) {}`
We'll have to do all the bracket matching to figure out which commas separate variable decls, and which are part of the type.

I'd really like to avoid that if possible, It requires a rather deep dive into parsing rust types, and is potentially fragile if new brackets ever get added to types (unlikely?).

Perhaps i'm missing some way we can avoid parsing these out though.
rust macro?

Reply:
Point taken: I hadn't considered all these cases! I do think that we don't have to worry too much about invalid syntax, provided that we can parse valid syntax :)

I wonder if we can do something fairly simple and get away with it. Let's take your example `fn parse<'a>(s:&'a str, (x, y) : (u32, u32)) {}`. If I parse everything after the lifetime, I think I can get away with the following algorithm:

1) Everything to the left of a `:` must be a name (or some brackets etc. which we could even ignore).
2) Everything to the right of a `:` must be types, upto the next `,` or `)`.
3) We have to count brackets (`<{(` etc.), but only in a simple way.

I don't think we necessarily have to check that the user is using valid Rust syntax. For example I'd be fine with allowing this to squeeze through `<(x>)` i.e. getting the closing brackets "the wrong way around" at the moment.

Put another way: I'd rather we get a nice syntax that we can live with long-term and maybe have a parser which is a little sloppy (in the sense of accepting some invalid things) since we can improve that over time. But once we've chosen a syntax, it becomes painful for users to change in the future.

Reply:
Yeah, it doesn't seem like that would be too painful, i'll try and work on it in the morning.

Reply:
Thanks :)

Reply:
Alright, so that is done.